rust/arrow/src/array/array.rs

```diff
@@ -1647,6 +1669,172 @@ impl From<(Vec<(Field, ArrayRef)>, Buffer, usize)> for StructArray {
+{
+    type Item = Option<T>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let i = self.i;
+        if i >= self.keys.len() {
+            None
+        } else if self.data.is_null(i) {
+            self.i += 1;
+            Some(None)
+        } else {
+            self.i += 1;
+            Some(Some(self.keys[i].clone()))
+        }
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        (self.keys.len(), Some(self.keys.len()))
+    }
+
+    fn nth(&mut self, n: usize) -> Option<Self::Item> {
+        let i = self.i;
+        if i + n >= self.keys.len() {
+            self.i = self.keys.len();
+            None
+        } else if self.data.is_null(i + n) {
+            self.i += n + 1;
+            Some(None)
+        } else {
+            self.i += n + 1;
+            Some(Some(self.keys[i + n].clone()))
+        }
+    }
+}
+
+impl<'a, K: ArrowPrimitiveType> DictionaryArray<K> {
+    /// Returns a slice of the keys (offsets into the values array).
+    /// Note that if nulls exist, you must check them.
+    pub fn keys(&self) -> &[K::Native] {
+        unsafe {
+            std::slice::from_raw_parts(
+                self.raw_values.get().offset(self.data.offset() as isize),
+                self.len(),
+            )
+        }
+    }
+
+    /// Returns an index into the values array or None if null.
+    pub fn key(&self, i: usize) -> Option<K::Native> {
+        if self.data.is_null(i) {
+            None
+        } else {
+            Some(self.keys()[i])
+        }
+    }
+
+    /// Return an iterator to the keys of this dictionary.
+    pub fn iter_keys(&'a self) -> NullableIter<'a, K::Native> {
+        NullableIter::<'a, K::Native> {
+            data: &self.data,
+            keys: self.keys(),
+            i: 0,
+        }
+    }
+
+    /// Returns an `ArrayRef` to the dictionary values.
+    pub fn values(&self) -> ArrayRef {
+        self.values.clone()
+    }
+
+    /// Returns a clone of the value type of this list.
+    pub fn value_type(&self) -> DataType {
+        self.values.data().data_type().clone()
+    }
+
+    /// The length of the dictionary is the length of the keys array.
+    pub fn len(&self) -> usize {
+        self.data.len()
+    }
+
+    pub fn is_ordered(&self) -> bool {
+        self.is_ordered
+    }
+}
+
+/// Constructs a `DictionaryArray` from an array data reference.
+impl<T: ArrowPrimitiveType> From<ArrayDataRef> for DictionaryArray<T> {
+    fn from(data: ArrayDataRef) -> Self {
+        assert_eq!(
+            data.buffers().len(),
+            1,
+            "DictionaryArray data should contain a single buffer only (keys)."
+        );
+        assert_eq!(
+            data.child_data().len(),
+            1,
+            "DictionaryArray should contain a single child array (values)."
+        );
+
+        let raw_values = data.buffers()[0].raw_data();
+        let dtype: &DataType = data.data_type();
+        let values = make_array(data.child_data()[0].clone());
```

Comment:
Am I correct that the DictionaryArray keeps 2 copies of the values data, one in `self.data.child_data()[0]` and the other in `values: Arc<Array>`? I noticed this while trying to get the keys as an arrow array

Reply:
I copied this code from the ListArray. I'm hoping (fervently) that it only clones the Arc and hence adds 1 to the ref count. This should take about 400ish cyles on Intel.

Reply:
You're correct, I looked at the `ListArray` implementation